In [1]:
import requests
import pandas as pd
import os
import re
from pandas.io.json import json_normalize
import json

爬取歌单名称：华语评论1W＋

歌单共计2000余首华语评论1000+的歌曲，涵盖了较老的歌和较新的歌，包罗万象

歌单作者id：133511246

歌单id：728958078

API:/playlist/track/all?id=728958078&limit=50&offset=0

首先先获取歌单里面所有歌曲的id

In [2]:
url = "http://127.0.0.1:3000/playlist/track/all?id=728958078&limit=10&offset=0"
response = requests.request("GET", url).json()

In [3]:
response

{'songs': [{'name': '危险派对',
   'id': 1877996649,
   'pst': 0,
   't': 0,
   'ar': [{'id': 12236125, 'name': '王以太', 'tns': [], 'alias': []},
    {'id': 1172023, 'name': '刘至佳', 'tns': [], 'alias': []}],
   'alia': [],
   'pop': 100,
   'st': 0,
   'rt': '',
   'fee': 8,
   'v': 27,
   'crbt': None,
   'cf': '',
   'al': {'id': 133202644,
    'name': '危险派对',
    'picUrl': 'https://p1.music.126.net/GpnLproqUUyc4xmYKpRFcQ==/109951166516282895.jpg',
    'tns': [],
    'pic_str': '109951166516282895',
    'pic': 109951166516282900},
   'dt': 180571,
   'h': {'br': 320000, 'fid': 0, 'size': 7225005, 'vd': -57704, 'sr': 48000},
   'm': {'br': 192000, 'fid': 0, 'size': 4335021, 'vd': -55129, 'sr': 48000},
   'l': {'br': 128000, 'fid': 0, 'size': 2890029, 'vd': -53450, 'sr': 48000},
   'sq': {'br': 977692, 'fid': 0, 'size': 22067908, 'vd': -57699, 'sr': 48000},
   'hr': {'br': 1747706,
    'fid': 0,
    'size': 39448232,
    'vd': -57699,
    'sr': 48000},
   'a': None,
   'cd': '01',
   'no': 1,

In [4]:
def GetSongs(offset):
    url_ = "http://127.0.0.1:3000/playlist/track/all?id=728958078&limit=1000&offset=" + str(offset)
    data_ = requests.request("GET", url_).json()
    df_ = pd.json_normalize(data_['songs'])
    return df_['id']

In [5]:
list1 = GetSongs(0).tolist()
list2 = GetSongs(1000).tolist()
list3 = GetSongs(2000).tolist()
list1.extend(list2)
list1.extend(list3)

In [6]:
len(list1)

2213

In [25]:
len(list1[1000:2000])

1000

经过复核歌单确实有2213首歌曲。接下来按照表单`list1`来爬取这中间每一首歌的热度最高的五条评论。

API:/comment/new?type=0&id=1877996649&sortType=2&pageSize=5

In [11]:
# 获取歌曲前五条热评
def GetComments(id):
    url_ = "http://127.0.0.1:3000/comment/new?type=0&id=" + str(id) + "&sortType=2&pageSize=5"
    data_ = requests.request("GET", url_).json()
    COLUMNS = ["user.userId","commentId", "user.nickname", "content", "timeStr", "likedCount", 'user.isHug', 'replyCount', 'user.authStatus']
    df_ = pd.json_normalize(data_['data']["comments"])
    return df_[COLUMNS]

In [12]:
def get_all_comments(song_list):
    ans = GetComments(song_list[0])
    for i in range(1,len(song_list)):
        new_df = GetComments(song_list[i])
        ans = pd.concat([ans, new_df])
    return ans

In [13]:
hot_comments = get_all_comments(list1)

In [32]:
hot_comments.to_csv("hot_comments_auth.csv",encoding='utf_8_sig')

In [15]:
hot_comments.head(50)

user.userId   commentId    user.nickname  \
0   3291743001  5428603245        玛卡巴卡在阿巴阿巴   
1    267627817  5427142964          皮崽-承蒙厚爱   
2    428956762  5419339232             保久伽壹   
3   1984646294  5446242912             阿泽潮音   
4    287711061  5420069620            等X的新歌   
0     47774029     9074483          胖了一整个盛夏   
1     67404460   128604071           小白不会很白   
2    128843665   339992770             若凤求凰   
3    122700987   135827338           你我皆凡人_   
4   4022472656  5627762154             圈钱公主   
0    474348590  5324462466           夏日入侵企画   
1   1338726532  5324547833              灰鸿啊   
2     75685933  5324532998              沐梓忻   
3   1831790043  5346915564          小蛇仙天天开心   
4    542077732  5327159696          蛙蛙_7777   
0     80303269  3392540986              谢白点   
1     84242841  3392524168             小含含_   
2    494841338  3394884358             少年柴柴   
3     80303269  3392562085              谢白点   
4    284313777  3394695262      1MWATCHINGU   
0   7811353507  5695587968            帐号已注销   
1   1494202911  5677351824             她的现任   
2   1777082129  5677099953           永远的二零三   
3   1752383264  5678537123            毛卜易yi   
4   7809907302  5703687964     TheCrossingg   
0     18590405     6931728           YSHY1丶   
1     18545834     7422591  突然发现手机里只有一个音乐软件   
2     18986628     6769250             雪源_L   
3     69247130    27066316          为往圣而继绝学   
4    121953443    60314861            追忆那年秋   
0    415974922   716926950            帐号已注销   
1      2059439     7962709             一粒馒头   
2     39441587     5929860              乖心儿   
3     31238729     7537077              _缄箴   
4     58978292    43988550            文娜mwn   
0   1755978650  5165621062         秋刀鱼_1996   
1   3864169274  5165589574          郝嘉璐是大魔王   
2    486274527  3587207410            ZzHpf   
3    603979509  5224179540          妳還是失约了-   
4   1944615402  5178591549           尤咏慈__-   
0     30905679     6998451           灬死亡之翼灬   
1      3360235     9403158     Jasmine_Yuqi   
2     31735416     4847969           Anno33   
3    273182140   202798524      ZdiscoveryZ   
4      3592708     9573458            藝術館假期   
0    277962003   191429332             风pdd   
1     41127627     5641383     Double___two   
2    124063058   131358133             黑白薄荷   
3     39354530     6750215         wongtian   
4      9380452     8780064              幻梦_   

                                             content     timeStr  likedCount  \
0                     不行啊！这歌也太让人快乐了！洗澡的时候我直接扭成龙卷风！！！  2021-10-01       95720   
1                      这派对也太危险了，男孩子在外面玩一定要好好保护自己[惊恐]  2021-09-29       68632   
2                                  “右手托住你的心跳” 怎么……托😬  2021-09-17       61614   
3                    你说你大姨妈走了 让我晚上去你家 宝  守灵这活哥真干不了啊   2021-10-27       44296   
4                                     托住你的心跳这句可太高情商了  2021-09-18       37329   
0                         开头金润吉的那几句简直好听炸！！！！！！！！！！\n  2015-01-20       27464   
1                           阿里郎这个名字起的俗，不然能火[大哭]是这样的吗  2016-03-12       18959   
2  有人问我思念到极致是什么感觉。 我曾经发了句晚安给她，一晚上醒来六、七次看手机信息。  就是...  2017-03-26       16671   
3                               今天2016年03月30号。这首歌会火！  2016-03-30       13079   
4                                      如果你刷到我 愿你嫁给爱情  2022-04-03        9039   
0  经常在夏天深夜出门散心。\n平缓流淌的街灯和温暖的晚风，被熟悉的气味和环境唤醒回忆，很容易让...  2021-06-05       19375   
1  乐队的第一部长篇「夏日黄昏三部曲」的终章。\n讲完了一段感情从「挽回」到「告别」再到「释然」...  2021-06-05       11947   
2                                 摇滚人是燃着温柔之火的极端浪漫主义者  2021-06-05        9622   
3                                        分享欲就是赤裸裸的告白  2021-07-02        6361   
4                                        夏天的歌还是得交给夏企  2021-06-08        4988   
0  这个周末一起早起打扫卫生吧，我愿意洗床单，你只需要在旁边拆好你的快递，收拾干净后洗个香香的热...  2020-07-26        8804   
1  三十而粒，你写世界正中。\n在足够漫长的时光中解构出自己生活的艺术，拥有过锋利的漂亮，要温柔...  2020-07-26        8768   
2                                      感觉她又和小女孩谈恋爱了。  2020-07-28        7657   
3          上次粉丝说想和二十岁的陈粒谈恋爱，她说“师傅 我现在也还可以的。” 嗯 真的很可以  2020-07-26        6393   
4    

In [16]:
# 获取歌曲前15条最新评论
def GetNewComments(id):
    url_ = "http://127.0.0.1:3000/comment/new?type=0&id=" + str(id) + "&sortType=3&pageSize=15"
    data_ = requests.request("GET", url_).json()
    COLUMNS = ["user.userId","commentId", "user.nickname", "content", "timeStr", "likedCount", 'user.isHug', 'replyCount']
    df_ = pd.json_normalize(data_['data']["comments"])
    return df_[COLUMNS]

In [17]:
def get_all_new_comments(song_list):
    ans = GetNewComments(song_list[0])
    for i in range(1,len(song_list)):
        new_df = GetNewComments(song_list[i])
        ans = pd.concat([ans, new_df])
    return ans

In [26]:
new_comments_1 = get_all_new_comments(list1[0:1000])

In [27]:
new_comments_2 = get_all_new_comments(list1[1000:])

In [29]:
new_comments = pd.concat([new_comments_1, new_comments_2])

In [30]:
new_comments

user.userId   commentId user.nickname                     content timeStr  \
0    1894471710  5905868958        银雾1906                           6    3分钟前   
1    6413620741  5905772200       摆烂的旅行者啊        有没有看过散兵和流浪者一起跳的。越看越涩   09:32   
2    3811137102  5905769848        理想超sky            让我想起了优拉跳舞的画面[18]   09:25   
3    8558486705  5905764073          许雨-1       是我们没有挺过去 别怪时间 更别说不适合.   09:11   
4    1690241045  5905747468        就把他打死吧                       音乐合伙人   08:20   
..          ...         ...           ...                         ...     ...   
10   8371314927  5902779771     _L-yyyyyy  困住我的是内心，是思念，是徘徊，是你说你从来不爱我。     6天前   
11   8371314927  5902779735     _L-yyyyyy         没有在为情所困 只是觉得身边的人都不诚     6天前   
12   8371314927  5902776250     _L-yyyyyy                 我试图挽留过，但没用。     6天前   
13   8371314927  5902775604     _L-yyyyyy                 我试图挽留过，但没用。     6天前   
14   1485564162  5902636530       山河_6QE8                  现在终于听懂这首歌了     6天前   

    likedCount  user.isHug  replyCount  
0            0       False           0  
1            0       False           0  
2            1       False           0  
3            0       False           0  
4            0       False           0  
..         ...         ...         ...  
10           0       False           0  
11           0       False           0  
12           0       False           0  
13           0       False           0  
14           0       False           0  

[33195 rows x 8 columns]

In [31]:
new_comments.to_csv("new_comments.csv",encoding='utf_8_sig')